File: scripts/ha-ws-client/src/handlers/registry.ts
/**
 * Registry command handlers for entities, devices, and areas.
 * @module handlers/registry
 */

import { sendMessage } from '../client.js';
import { getOutputConfig, output, outputList, outputMessage } from '../output.js';
import type { AreaEntry, CommandContext, DeviceEntry, EntityEntry } from '../types.js';

/**
 * List or search the entity registry.
 * Shows entity metadata including platform and disabled status.
 *
 * @param ctx - Command context with WebSocket and arguments
 *
 * @example
 * ```bash
 * npx tsx ha-ws-client.ts entities
 * npx tsx ha-ws-client.ts entities "kitchen"
 * npx tsx ha-ws-client.ts entities "light.*"
 * ```
 */
export async function handleEntities(ctx: CommandContext): Promise<void> {
  const pattern = ctx.args[1];
  const result = await sendMessage<EntityEntry[]>(ctx.ws, 'config/entity_registry/list');

  let filtered = result;
  if (pattern) {
    const regex = new RegExp(pattern, 'i');
    filtered = result.filter(
      (e) =>
        regex.test(e.entity_id) || regex.test(e.name ?? '') || regex.test(e.original_name ?? '')
    );
  }

  const { format, maxItems } = getOutputConfig();
  const limit = maxItems > 0 ? maxItems : 30;

  if (format === 'json') {
    output(filtered.slice(0, limit), { command: 'entities', count: filtered.length });
    return;
  }

  outputList(filtered.slice(0, limit), {
    title: `Entity registry${pattern ? ` matching "${pattern}"` : ''}: ${filtered.length} entities`,
    command: 'entities',
    itemFormatter: (e) => {
      const name = e.name ?? e.original_name ?? '';
      const platform = e.platform ?? '';
      const disabled = e.disabled_by ? ' [DISABLED]' : '';
      if (format === 'compact') {
        return `${e.entity_id}${disabled}${name ? ` (${name})` : ''}`;
      }
      let output = `  ${e.entity_id}${disabled}`;
      if (name) output += `\n    name: ${name}`;
      if (platform) output += `\n    platform: ${platform}`;
      return output;
    },
  });
  if (filtered.length > limit) {
    outputMessage(`\n  ... and ${filtered.length - limit} more`);
  }
}

/**
 * List or search the device registry.
 * Shows device metadata including manufacturer, model, and area.
 *
 * @param ctx - Command context with WebSocket and arguments
 *
 * @example
 * ```bash
 * npx tsx ha-ws-client.ts devices
 * npx tsx ha-ws-client.ts devices "philips"
 * npx tsx ha-ws-client.ts devices "living_room"
 * ```
 */
export async function handleDevices(ctx: CommandContext): Promise<void> {
  const pattern = ctx.args[1];
  const result = await sendMessage<DeviceEntry[]>(ctx.ws, 'config/device_registry/list');

  let filtered = result;
  if (pattern) {
    const regex = new RegExp(pattern, 'i');
    filtered = result.filter(
      (d) =>
        regex.test(d.name ?? '') ||
        regex.test(d.name_by_user ?? '') ||
        regex.test(d.manufacturer ?? '') ||
        regex.test(d.model ?? '') ||
        regex.test(d.area_id ?? '')
    );
  }

  const { format, maxItems } = getOutputConfig();
  const limit = maxItems > 0 ? maxItems : 20;

  if (format === 'json') {
    output(filtered.slice(0, limit), { command: 'devices', count: filtered.length });
    return;
  }

  outputList(filtered.slice(0, limit), {
    title: `Device registry${pattern ? ` matching "${pattern}"` : ''}: ${filtered.length} devices`,
    command: 'devices',
    itemFormatter: (d) => {
      const name = d.name ?? d.name_by_user ?? 'Unnamed';
      if (format === 'compact') {
        const parts = [name, d.manufacturer, d.model, d.area_id].filter(Boolean);
        return parts.join(' | ');
      }
      let output = `  ${name}\n    id: ${d.id}`;
      if (d.manufacturer) output += `\n    manufacturer: ${d.manufacturer}`;
      if (d.model) output += `\n    model: ${d.model}`;
      if (d.area_id) output += `\n    area: ${d.area_id}`;
      return output;
    },
  });
  if (filtered.length > limit) {
    outputMessage(`\n  ... and ${filtered.length - limit} more`);
  }
}

/**
 * List all areas in the area registry.
 * Shows area names, IDs, and any aliases.
 *
 * @param ctx - Command context with WebSocket and arguments
 *
 * @example
 * ```bash
 * npx tsx ha-ws-client.ts areas
 * ```
 */
export async function handleAreas(ctx: CommandContext): Promise<void> {
  const result = await sendMessage<AreaEntry[]>(ctx.ws, 'config/area_registry/list');
  const { format } = getOutputConfig();

  if (format === 'json') {
    output(result, { command: 'areas', count: result.length });
    return;
  }

  outputList(result, {
    title: 'Areas',
    command: 'areas',
    itemFormatter: (a) => {
      if (format === 'compact') {
        const aliases = a.aliases && a.aliases.length > 0 ? ` [${a.aliases.join(', ')}]` : '';
        return `${a.area_id}: ${a.name}${aliases}`;
      }
      let output = `  ${a.name} (${a.area_id})`;
      if (a.aliases && a.aliases.length > 0) {
        output += `\n    aliases: ${a.aliases.join(', ')}`;
      }
      return output;
    },
  });
}

File: scripts/ha-ws-client-go/internal/handlers/registry.go
package handlers

import (
	"fmt"
	"regexp"
	"slices"
	"strings"

	"github.com/home-assistant-blueprints/ha-ws-client-go/internal/client"
	"github.com/home-assistant-blueprints/ha-ws-client-go/internal/output"
	"github.com/home-assistant-blueprints/ha-ws-client-go/internal/types"
)

func init() {
	// Register entity/device/area registry commands
	RegisterAll(
		Cmd("entities", "List/search entity registry", "[pattern]", "registry", HandleEntities),
		Cmd("devices", "List/search device registry", "[pattern]", "registry", HandleDevices),
		Cmd("areas", "List all areas", "", "registry", HandleAreas),
	)
}

// RegistryConfig holds the configuration for a filtered registry handler.
type RegistryConfig[T any] struct {
	// MessageType is the WebSocket message type to fetch data (e.g., "config/entity_registry/list").
	MessageType string
	// Title is the list title shown in output.
	Title string
	// Command is the command name for JSON output.
	Command string
	// MatchFields returns the string fields to match against the pattern for a given item.
	MatchFields func(item T) []string
	// Formatter formats an item for display output.
	Formatter func(item T, index int) string
}

// FilteredRegistryHandler creates a handler that fetches registry data,
// optionally filters by pattern, and outputs the results.
// This is a generic helper that unifies the pattern-filtering logic used by
// entities, devices, and similar registry commands.
func FilteredRegistryHandler[T any](cfg RegistryConfig[T]) Handler {
	return func(ctx *Context) error {
		re := ctx.Config.Pattern

		items, err := client.SendMessageTyped[[]T](ctx.Client, cfg.MessageType, nil)
		if err != nil {
			return err
		}

		// Filter by pattern if provided
		if re != nil {
			items = filterByPattern(items, re, cfg.MatchFields)
		}

		output.List(items,
			output.ListTitle[T](cfg.Title),
			output.ListCommand[T](cfg.Command),
			output.ListFormatter(cfg.Formatter),
		)
		return nil
	}
}

// filterByPattern filters a slice of items, keeping only those where at least
// one of the match fields matches the given pattern.
func filterByPattern[T any](items []T, re *regexp.Regexp, matchFields func(T) []string) []T {
	var filtered []T
	for _, item := range items {
		if slices.ContainsFunc(matchFields(item), re.MatchString) {
			filtered = append(filtered, item)
		}
	}
	return filtered
}

// HandleEntities lists or searches the entity registry.
// Wrapped with: WithPattern(1)
var HandleEntities = Apply(
	WithPattern(1),
	FilteredRegistryHandler(RegistryConfig[types.EntityEntry]{
		MessageType: "config/entity_registry/list",
		Title:       "Entity registry",
		Command:     "entities",
		MatchFields: func(e types.EntityEntry) []string {
			return []string{e.EntityID, e.Name, e.OriginalName}
		},
		Formatter: func(e types.EntityEntry, _ int) string {
			name := e.Name
			if name == "" {
				name = e.OriginalName
			}
			disabled := ""
			if e.DisabledBy != "" {
				disabled = " [disabled]"
			}
			if output.IsCompact() {
				return fmt.Sprintf("%s (%s)%s", e.EntityID, name, disabled)
			}
			return fmt.Sprintf("%s\n  Name: %s\n  Platform: %s%s", e.EntityID, name, e.Platform, disabled)
		},
	}),
)

// HandleDevices lists or searches the device registry.
// Wrapped with: WithPattern(1)
var HandleDevices = Apply(
	WithPattern(1),
	FilteredRegistryHandler(RegistryConfig[types.DeviceEntry]{
		MessageType: "config/device_registry/list",
		Title:       "Device registry",
		Command:     "devices",
		MatchFields: func(d types.DeviceEntry) []string {
			name := d.Name
			if d.NameByUser != "" {
				name = d.NameByUser
			}
			return []string{d.ID, name, d.Manufacturer, d.Model}
		},
		Formatter: func(d types.DeviceEntry, _ int) string {
			name := d.Name
			if d.NameByUser != "" {
				name = d.NameByUser
			}
			if output.IsCompact() {
				return fmt.Sprintf("%s: %s (%s %s)", d.ID[:8], name, d.Manufacturer, d.Model)
			}
			return fmt.Sprintf("%s\n  Name: %s\n  Manufacturer: %s\n  Model: %s\n  Area: %s",
				d.ID, name, d.Manufacturer, d.Model, d.AreaID)
		},
	}),
)

// HandleAreas lists all areas.
func HandleAreas(ctx *Context) error {
	areas, err := client.SendMessageTyped[[]types.AreaEntry](ctx.Client, "config/area_registry/list", nil)
	if err != nil {
		return err
	}

	output.List(areas,
		output.ListTitle[types.AreaEntry]("Areas"),
		output.ListCommand[types.AreaEntry]("areas"),
		output.ListFormatter(func(a types.AreaEntry, _ int) string {
			aliases := ""
			if len(a.Aliases) > 0 {
				aliases = fmt.Sprintf(" (aliases: %s)", strings.Join(a.Aliases, ", "))
			}
			if output.IsCompact() {
				return fmt.Sprintf("%s: %s%s", a.AreaID, a.Name, aliases)
			}
			return fmt.Sprintf("%s: %s%s", a.AreaID, a.Name, aliases)
		}),
	)
	return nil
}
